taichi/transforms/redundancy_in_loop_elimination.cpp

```diff
@@ -0,0 +1,146 @@
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+#include "taichi/system/profiler.h"
+
+#include <stack>
+
+TLANG_NAMESPACE_BEGIN
+
+//#define LOOP_INVARIANT_ONLY_ALLOWING_LOOPS
+
+class RedundancyInLoopElimination : public IRVisitor {
+ public:
+  bool modified = false;
+
+  std::stack<Block *> loop_blocks;
+
+  DelayedIRModifier modifier;
+
+  RedundancyInLoopElimination() {
+    allow_undefined_visitor = true;
+  }
+
+  bool visit_arithmetic(Stmt *stmt) {
+#ifdef LOOP_INVARIANT_ONLY_ALLOWING_LOOPS
+    if (loop_blocks.size() <= 1 || stmt->parent != loop_blocks.top())
+      return false;
+#else
+    if (loop_blocks.size() <= 1)
+      return false;
+#endif
+
+    bool can_be_moved = true;
+
+    Block *current_scope = stmt->parent;
+
+    for (Stmt *operand : stmt->get_operands()) {
+      if (operand->parent == current_scope) {
+        // Has an input that is in the current scope
+        // Operand can either be in the current scope or a outer scope
+        can_be_moved = false;
+        break;
+      }
+    }
+
+    if (can_be_moved) {
+      TI_TRACE("Stmt can be moved: {} {}", stmt->name(), typeid(*stmt).name());
+    }
+
+    return can_be_moved;
+  }
+
+  void visit(BinaryOpStmt *stmt) override {
+    if (visit_arithmetic(stmt)) {
+      auto replacement = stmt->clone();
+      stmt->replace_with(replacement.get());
+
+      modifier.insert_before(stmt->parent->parent_stmt, std::move(replacement));
+      modifier.erase(stmt);
+
+      modified = true;
+    }
+  }
+
+  void visit(UnaryOpStmt *stmt) override {
+    if (visit_arithmetic(stmt)) {
+      auto replacement = stmt->clone();
+      stmt->replace_with(replacement.get());
+
+      modifier.insert_before(stmt->parent->parent_stmt, std::move(replacement));
+      modifier.erase(stmt);
+
+      modified = true;
+    }
+  }
+
+  void visit(Block *stmt_list) override {
+    for (auto &stmt : stmt_list->statements)
+      stmt->accept(this);
+  }
+
+  void visit_loop(Block *body) {
+    loop_blocks.push(body);
+
+    body->accept(this);
+
+    loop_blocks.pop();
+  }
+  void visit(RangeForStmt *stmt) override {
+    visit_loop(stmt->body.get());
+  }
+
+  void visit(StructForStmt *stmt) override {
+    visit_loop(stmt->body.get());
+  }
+
+  void visit(WhileStmt *stmt) override {
+    visit_loop(stmt->body.get());
+  }
+
+  void visit(OffloadedStmt *stmt) override {
+    if (stmt->tls_prologue)
+      stmt->tls_prologue->accept(this);
+
+    if (stmt->bls_prologue)
+      stmt->bls_prologue->accept(this);
+
+    if (stmt->body)
+      stmt->body->accept(this);
+
+    if (stmt->bls_epilogue)
+      stmt->bls_epilogue->accept(this);
+
+    if (stmt->tls_epilogue)
+      stmt->tls_epilogue->accept(this);
+  }
+
+  void visit(IfStmt *if_stmt) override {
+    if (if_stmt->true_statements)
+      if_stmt->true_statements->accept(this);
+    if (if_stmt->false_statements)
+      if_stmt->false_statements->accept(this);
+  }
+
+  static bool run(IRNode *node) {
```

Comment:
Please modify this function like https://github.com/taichi-dev/taichi/blob/6586d6b53a2d39e6b274c4ec952215adb71896be/taichi/transforms/alg_simp.cpp#L316-L327 so that the return value is well-defined.